File: dotnet/path_traversal/FileDownload.cs
using System.IO;
using Microsoft.AspNetCore.Mvc;

[ApiController]
[Route("[controller]")]
public class FileDownloadController : ControllerBase
{
    private const string BaseDir = "/var/www/files";

    // VULN 1: System.IO.File.ReadAllText with string concatenation - no path sanitization
    [HttpGet("read")]
    public IActionResult ReadUserFile(string file)
    {
        string path = BaseDir + "/" + file;
        string content = System.IO.File.ReadAllText(path);
        return Ok(content);
    }

    // VULN 2: PhysicalFile with user-controlled path - allows ../../etc/passwd
    [HttpGet("download")]
    public IActionResult DownloadFile(string file)
    {
        string filepath = Path.Combine(BaseDir, file);
        return PhysicalFile(filepath, "application/octet-stream");
    }

    // VULN 3: FileStream with string interpolation - stream arbitrary files
    [HttpGet("export")]
    public IActionResult ExportReport(string name)
    {
        string fullPath = Path.Combine("/reports/output", name);
        byte[] data = System.IO.File.ReadAllBytes(fullPath);
        return File(data, "application/octet-stream");
    }
}

File: java/path_traversal/FileDownload.java
import jakarta.servlet.http.*;
import java.io.*;
import java.nio.file.*;

public class FileDownload extends HttpServlet {
    private static final String BASE_DIR = "/var/www/files";

    // VULN 1: Files.readAllBytes with string concatenation - no path sanitization
    protected void doGetRead(HttpServletRequest req, HttpServletResponse resp) throws IOException {
        String filename = req.getParameter("file");
        String path = BASE_DIR + "/" + filename;
        byte[] data = Files.readAllBytes(Paths.get(path));
        resp.getOutputStream().write(data);
    }

    // VULN 2: FileInputStream with Path.resolve - allows ../../etc/passwd
    protected void doGetDownload(HttpServletRequest req, HttpServletResponse resp) throws IOException {
        String filename = req.getParameter("file");
        Path filepath = Paths.get(BASE_DIR).resolve(filename);
        byte[] data = Files.readAllBytes(filepath);
        resp.getOutputStream().write(data);
    }

    // VULN 3: new File() with user-controlled path - stream arbitrary files
    protected void doGetExport(HttpServletRequest req, HttpServletResponse resp) throws IOException {
        String reportName = req.getParameter("name");
        File file = new File("/reports/output/" + reportName);
        FileInputStream fis = new FileInputStream(file);
        fis.transferTo(resp.getOutputStream());
        fis.close();
    }
}
